internal/oci/runtime_oci.go

```diff
@@ -1402,7 +1415,19 @@ func (r *runtimeOCI) runtimeCmd(args ...string) (string, error) {
 
 	err := cmd.Run()
 	if err != nil {
-		return "", fmt.Errorf("`%v %v` failed: %v %v: %w", r.handler.RuntimePath, strings.Join(runtimeArgs, " "), stderr.String(), stdout.String(), err)
+		stderrStr := stderr.String()
+		var runtimeErr error
+		switch {
+		case strings.Contains(stderrStr, "no such process"):
+			runtimeErr = ErrNoSuchProcess
+		case strings.Contains(stderrStr, "container not running"):
+			runtimeErr = ErrNotRunning
+		case strings.Contains(stderrStr, "invalid process"):
+			runtimeErr = ErrInvalidProcess
+		default:
+			runtimeErr = err
+		}
+		return "", fmt.Errorf("`%v %v` failed: %v %v: %w", r.handler.RuntimePath, strings.Join(runtimeArgs, " "), stderr.String(), stdout.String(), runtimeErr)
```

Comment:
We can probably keep it simple, for example:

```go
err := cmd.Run()
if err != nil {
	// TODO: Remove this function when we have standardized exit codes or
	// error reporting from container runtimes. Current approach is a
	// temporary workaround based on error string matching.
	errString := stderr.String()
	switch {
	// crun, for most of the commands.
	case strings.Contains(errString, "no such process"):
		fallthrough
	// runc, for most of the commands.
	case strings.Contains(errString, "container not running"):
		fallthrough
	// runc, on a rare occassion.
	case strings.Contains(errString, "invalid process"):
		err = ErrNotFound
	}
	return "", fmt.Errorf("`%v %v` failed: %v %v: %w", r.handler.RuntimePath, strings.Join(runtimeArgs, " "), stderr.String(), stdout.String(), err)
}
```

The kill loop would get (same as the background goroutine):

```go
if _, err := r.runtimeCmd("kill", c.ID(), "KILL"); err != nil {
	log.Debugf(ctx, "Error while killing container %s: %v", c.ID(), err)
	if !errors.Is(err, ErrNotFound) {
		log.Errorf(ctx, "Killing container %v failed: %v", c.ID(), err)
	}
}
```

The `log.Debugf()` is there to ensure visibility in case something is failing and we do not eat the error up. Can be optional, of course.

Reply:
Yeah, this is much better.